laplace/curvature/curvlinops.py

```diff
@@ -0,0 +1,123 @@
+import torch
+import numpy as np
+
+import curvlinops as cvls
+
+from laplace.curvature import CurvatureInterface, GGNInterface, EFInterface
+from laplace.utils import Kron
+
+from collections import UserDict
+
+from typing import *
+
+
+class CurvlinopsInterface(CurvatureInterface):
+    """Interface for Curvlinops backend. <https://github.com/f-dangel/curvlinops>
+    """
+    def __init__(self, model, likelihood, last_layer=False, subnetwork_indices=None):
+        super().__init__(model, likelihood, last_layer, subnetwork_indices)
+
+    @property
+    def _kron_fisher_type(self):
+        raise NotImplementedError
+
+    @property
+    def _linop_context(self):
+        raise NotImplementedError
+
+    @staticmethod
+    def _rescale_kron_factors(kron, M, N):
+        # Renormalize Kronecker factor to sum up correctly over N data points with batches of M
+        # for M=N (full-batch) just M/N=1
+        for F in kron.kfacs:
+            if len(F) == 2:
+                F[1] *= M/N
+        return kron
+
+    def _get_kron_factors(self, linop, M):
+        kfacs = list()
+        for name, module in self._model.named_modules():
+            if name not in linop._mapping.keys():
+                continue
+
+            A = linop._input_covariances[name]
+            B = linop._gradient_covariances[name]
+
+            if hasattr(module, 'bias') and module.bias is not None:
+                # split up bias and weights
+                kfacs.append([B, A[:-1, :-1]])
+                kfacs.append([B * A[-1, -1] / M])
+            elif hasattr(module, 'weight'):
+                p, q = np.prod(B.shape), np.prod(A.shape)
+                if p == q == 1:
+                    kfacs.append([B * A])
+                else:
+                    kfacs.append([B, A])
+            else:
+                raise ValueError(f'Whats happening with {module}?')
+        return Kron(kfacs)
+
+    def kron(self, X, y, N, **kwargs):
+        linop = cvls.KFACLinearOperator(
+            self.model, self.lossfunc, self.params, [(X, y)],
+            fisher_type=self._kron_fisher_type,
+            loss_average=None,  # Since self.lossfunc is sum
+            separate_weight_and_bias=False
+        )
+        linop._compute_kfac()
+
+        kron = self._get_kron_factors(linop, len(y))
+        kron = self._rescale_kron_factors(kron, len(y), N)
+
+        loss = self.lossfunc(self.model(X), y)
+
+        return self.factor * loss.detach(), self.factor * kron
```

Comment:
@runame please check this implementation. I'm not sure how to proceed. Comparing to BackPACK's & ASDL's KFAC, I still get the wrong results. 

If you need a test script: <https://gist.github.com/wiseodd/d9445133635a5d1d01f5b4c6e387586c>